scipy/spatial/transform/rotation.py

```diff
@@ -2061,13 +2069,31 @@ def __init__(self, times, rotations):
                              "and {} timestamps.".format(
                                 len(rotations), times.shape[0]))
         self.times = times
-        self.timedelta = np.diff(times)
-
+        self.timedelta = np.concatenate((np.diff(times), [1]))
         if np.any(self.timedelta <= 0):
             raise ValueError("Times must be in strictly increasing order.")
 
-        self.rotations = rotations[:-1]
-        self.rotvecs = (self.rotations.inv() * rotations[1:]).as_rotvec()
+        starts = rotations.as_quat()
+        ends = np.roll(starts, -1, axis=0)
+        dots = np.einsum('ij,ij->i', starts, ends)
+        signs = 2 * (dots >= 0) - 1
+        ends *= signs[:, np.newaxis]
+
+        self.basis = []
+        self.omega = []
+        for i in range(len(starts)):
```

Comment:
Can we vectorize that loop?

Reply:
I have vectorized the angle calculation. The QR decomposition accepts square matrices only, so I can't vectorize that part.